compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -438,14 +516,234 @@ void distributeTransferWrites(vector::TransferWriteOp writeOp,
   }
 }
 
+static bool isLaneId(int dimType) {
+  return ((dimType == DimType::LaneIdX) || (dimType == DimType::LaneIdY) ||
+          (dimType == DimType::LaneIdZ));
+}
+
+static bool isVectorId(int dimType) {
+  return ((dimType == DimType::VecIdX) || (dimType == DimType::VecIdY) ||
+          (dimType == DimType::VecIdZ));
+}
+
+static int getLaneIdIndex(std::array<int, 4> &order) {
+  for (int i = 0; i < 4; i++) {
+    if (isLaneId(order[i])) return i;
+  }
+  return -1;
+}
+
+static int isSingleLaneIdReduced(std::array<int, 4> &order) {
+  int count{0};
+  for (int i = 0; i < 4; i++) {
+    if (isLaneId(order[i])) count++;
+  }
+  return count == 1;
+}
+
+static int getVecSizes(std::array<int, 4> &order, const Layout &layout) {
+  int size = 1;
+  for (int i = 0; i < 4; i++) {
+    if (isVectorId(i)) size *= layout.shape[i];
+  }
+  return size;
+}
+
+static Value createReductionOp(vector::CombiningKind combiningKind, Value lhs,
+                               Value rhs, Location loc, OpBuilder &rewriter) {
+  switch (combiningKind) {
+    case vector::CombiningKind::ADD:
+      return rewriter.create<arith::AddFOp>(loc, lhs, rhs);
+    case vector::CombiningKind::MAXF:
+      return rewriter.create<arith::MaxFOp>(loc, lhs, rhs);
+    default:
+      assert("Unsupported vector combining kind!");
+      break;
+  }
+  return Value{};
+}
+
+using Triplet = std::tuple<vector::MultiDimReductionOp, vector::BroadcastOp,
+                           vector::TransposeOp>;
+using bodyType = std::function<void(std::array<int, DimType::NumDims> &)>;
+
+template <std::size_t T>
+void iterate(int dimType, std::array<int, T> &order,
```

Comment:
If you use `ArrayRef<int>` can this not be a template?

It would be nice to add some doc to this function.

Reply:
Yes, that's right. I will change it to use ArrayRef and will add documentation as well.